File: packages/agent/agent/subagents/extractor/tools/read_pdf_pages.ts
import { openSealedInput } from "@clara-financas/auth/sealed-input";
import { getDb } from "@clara-financas/db";
import { documents } from "@clara-financas/db/schema/ledger";
import { forTenant } from "@clara-financas/db/tenant-scope";
import { and, eq } from "drizzle-orm";
import { defineTool } from "eve/tools";
import { z } from "zod";

import { requireTenantCaller } from "../../../lib/tenant";
import { extractPdfText, PdfPasswordRequiredError } from "../../../lib/pdf";

/**
 * Lê a camada de texto de um PDF, página a página.
 *
 * Por que texto e não imagem: a spec restringe a v1 a PDFs nativos digitais,
 * que saem dos emissores com camada de texto íntegra — não há OCR envolvido, e
 * o desafio real é interpretação de layout. Além disso, tools do eve devolvem
 * apenas texto ou JSON, nunca partes de arquivo, então enviar o PDF ao modelo
 * como imagem não é possível por esta via.
 *
 * O documento é buscado pelo id, sempre sob o escopo do tenant do chamador.
 */
export default defineTool({
  description:
    "Reads the text of an already-uploaded financial PDF, page by page. Use before extracting transactions. If the PDF is protected, supply the short-lived sealed password token.",
  inputSchema: z.object({
    documentId: z.string().min(1).describe("Id of the document the person uploaded."),
    passwordToken: z
      .string()
      .optional()
      .describe("Short-lived sealed password token, when protected. Never ask for plaintext."),
    fromPage: z.number().int().positive().optional().describe("First page (1-based)."),
    toPage: z.number().int().positive().optional().describe("Last page, inclusive."),
  }),
  async execute(input, ctx) {
    const { tenantId } = requireTenantCaller(ctx);
    let password: string | undefined;

    if (input.passwordToken) {
      const secret = process.env.AGENT_TOKEN_SECRET;
      if (!secret) throw new Error("AGENT_TOKEN_SECRET não configurado");

      try {
        const sealed = await openSealedInput(input.passwordToken, secret);
        if (sealed.tenantId !== tenantId || sealed.purpose !== "pdf_password") {
          return { error: "credencial_invalida" as const };
        }
        password = sealed.value;
      } catch {
        return { error: "credencial_invalida" as const };
      }
    }

    const document = await forTenant(
      tenantId,
      async (tx) => {
        const [row] = await tx
          .select()
          .from(documents)
          .where(and(eq(documents.id, input.documentId), eq(documents.tenantId, tenantId)))
          .limit(1);
        return row;
      },
      getDb(),
    );

    if (!document) {
      // Não distinguimos "não existe" de "é de outro tenant": a diferença
      // vazaria a existência de documentos alheios.
      return { error: "documento não encontrado" as const };
    }

    try {
      const pages = await extractPdfText(document.blobKey, {
        password,
        fromPage: input.fromPage,
        toPage: input.toPage,
      });

      return {
        documentId: document.id,
        filename: document.filename,
        issuer: document.issuer,
        totalPages: pages.totalPages,
        pages: pages.pages,
      };
    } catch (error) {
      if (error instanceof PdfPasswordRequiredError) {
        return {
          error: "senha_necessaria" as const,
          message:
            "This PDF is password-protected. Ask the person for the password before retrying — do not guess.",
        };
      }
      throw error;
    }
  },
});

File: services/agent/clara/tools/read_pdf_pages.py
"""Lê a camada de texto de um PDF, página a página — `@tool` de fronteira.

Porta de `subagents/extractor/tools/read_pdf_pages.ts`. O documento é
buscado pelo id, sempre sob o escopo do tenant do chamador (FR-001).
Nenhum texto extraído é persistido por esta tool — quem persiste é
`save_extraction`, e só o que o extrator decidiu reter (FR-008).

A verificação de token de senha selado (`sealed-input`, TypeScript) não é
portada nesta fase: a senha chega em texto simples como argumento da tool.
É uma redução de escopo deliberada — ver docs/ledger-python.md — não um
descuido; a fronteira onde ela deveria ser reintroduzida é aqui.
"""

from __future__ import annotations

from agno.run.base import RunContext
from agno.tools import tool
from sqlalchemy import select

from clara.db.models import Document
from clara.db.tenant_scope import for_tenant
from clara.tools.errors import not_found, tool_error
from clara.tools.pdf import PdfPasswordRequiredError, extract_pdf_text
from clara.tools.serialize import to_tool_result
from clara.tools.tenant import require_tenant_caller


@tool(
    name="read_pdf_pages",
    description=(
        "Reads the text of an already-uploaded financial PDF, page by page. "
        "Use before extracting transactions. If the PDF is protected, ask the "
        "person for the password first."
    ),
)
def read_pdf_pages(
    run_context: RunContext,
    document_id: str,
    password: str | None = None,
    from_page: int | None = None,
    to_page: int | None = None,
) -> dict:
    caller = require_tenant_caller(run_context)

    with for_tenant(caller.tenant_id) as session:
        document = session.execute(
            select(Document).where(
                Document.id == document_id, Document.tenant_id == caller.tenant_id
            )
        ).scalar_one_or_none()

    if document is None:
        # Não distinguimos "não existe" de "é de outro tenant": a diferença
        # vazaria a existência de documentos alheios.
        return to_tool_result(
            not_found(
                "documento_nao_encontrado", "Nenhum documento com esse id.",
                hint="Use o document_id exatamente como veio na requisição da coordenadora.",
            )
        )

    try:
        extracted = extract_pdf_text(
            document.blob_key, password=password, from_page=from_page, to_page=to_page
        )
    except PdfPasswordRequiredError:
        return to_tool_result(
            tool_error(
                "senha_necessaria", "Este PDF é protegido por senha.",
                hint="Peça a senha à pessoa e repita a chamada com password preenchido.",
            )
        )

    return to_tool_result(
        {
            "document_id": document.id,
            "filename": document.filename,
            "issuer": document.issuer,
            "total_pages": extracted.total_pages,
            "pages": [{"page": p.page, "text": p.text} for p in extracted.pages],
        }
    )
